### PR TITLE
sql task add

### PR DESCRIPTION
**Pull request by NagorniyAleksandr**
check pls

### CODE REVIEW

File: NagorniyAleksandr/SQLTask/src/main/java/university/container/PropertiesHolder.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public static final String PROPERTIES_FILE_PATH = "/jdbc.properties";
+
+    private static Properties properties = load();
+
+    private static Properties load() {
+        Properties properties = new Properties();
+        try {
+            properties.load(PropertiesHolder.class.getResourceAsStream(PROPERTIES_FILE_PATH));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/university/container/PropertiesHolder.java]: log info

File: NagorniyAleksandr/SQLTask/src/main/resources/jdbc.properties [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# MySQL Properties
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/resources/jdbc.properties:1]: copy jdbc.properties to jdbc.properties and put it to test/resources

define h2 in-memory database for testing

File: NagorniyAleksandr/SQLTask/src/main/java/university/container/PropertiesHolder.java (hunk 1 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/java/university/container/PropertiesHolder.java]: fix

File: NagorniyAleksandr/SQLTask/src/main/resources/jdbc.properties (hunk 2 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/resources/jdbc.properties:1]: done

Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/resources/jdbc.properties:1]: Done one more time) Test DataBase = H2.

File: NagorniyAleksandr/SQLTask/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+# JDBC
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/README.md]: Try to describe a business logic of your project 
and put instructions how to run 
e.g. step by step overview

File: NagorniyAleksandr/SQLTask/src/main/java/testRun.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import university.jdbc.DBConnector;
+import university.jdbc.DBConnectorImpl;
+import university.models.Subject;
+import university.service.Service;
+import university.service.ServiceImpl;
+
+import java.util.List;
+
+
+public class testRun {
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/testRun.java]: Java Naming, A class name starts form first big letter

File: NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+                             "LIMIT ? OFFSET ?")) {
+
+            preparedStatement.setInt(1, length);
+            preparedStatement.setInt(2, offset);
+
+            ResultSet resultSet = preparedStatement.executeQuery();
+
+            return getStudentsAsListFromResultSet(resultSet);
+        } catch (SQLException e) {
+            log.error(e.getMessage());
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java]: call logger once, just add additional info as argument

File: NagorniyAleksandr/SQLTask/src/main/java/university/dao/converter/ToObjectConverter.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ToObjectConverter {
+
+    public static Student getOneStudentFromResultSet(ResultSet resultSet) throws SQLException {
+        if (resultSet.next()) {
+            int studentId = resultSet.getInt(1);
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/university/dao/converter/ToObjectConverter.java]: Please remember, that your logic depends on right order of table's colums

if someone will change the view of result table you will get a runtime error

File: NagorniyAleksandr/SQLTask/src/main/java/university/models/Subject.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        sb.append(", name='").append(name).append('\'');
+        sb.append(", category=").append(category);
+        sb.append(", description='").append(description).append('\'');
+        sb.append('}');
+        return sb.toString();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/university/models/Subject.java:66]: if you have the id field, take only it while overriding equals and hashCode

File: NagorniyAleksandr/SQLTask/src/main/java/university/service/ServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+import org.apache.log4j.Logger;
+import university.dao.QueryCreator;
+import university.exceptions.AppDBException;
+import university.exceptions.InvalidQueryParameterException;
+import university.models.*;
+
+import java.util.List;
+
+public class ServiceImpl implements Service {
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/main/java/university/service/ServiceImpl.java:13]: Add validation with Regexp
Example:
```addStudent(Student)```

check fields value of the student object
name, age and so on

File: NagorniyAleksandr/SQLTask/src/test/java/university/dao/crud/CRUDQuerySubjectTest.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+import university.models.SubjectCategory;
+
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.hamcrest.core.IsNot.not;
+import static org.junit.Assert.*;
+
+public class CRUDQuerySubjectTest {
```
Comment (presly808) [inline NagorniyAleksandr/SQLTask/src/test/java/university/dao/crud/CRUDQuerySubjectTest.java]: Several tests repeat beforeClass logic
Create the parent test which will init the schema and data
Parent class will be inherited by child tests

File: NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java (hunk 5 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java]: Disagrees. I catch SQLException and wrap them in my exceptions. And I would like to see in the log 
1. Exception(SQL Exception) in the class XXX with message "YYY", 
2. Throw new Exception(My exception) from class XXX. 
Then from logs I'll be able track "from->where" my exceptions fly.

File: NagorniyAleksandr/SQLTask/src/main/java/testRun.java (hunk 4 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/java/testRun.java]: fixed

File: NagorniyAleksandr/SQLTask/src/main/java/university/models/Subject.java (hunk 7 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/java/university/models/Subject.java:66]: fixed

File: NagorniyAleksandr/SQLTask/src/test/java/university/dao/crud/CRUDQuerySubjectTest.java (hunk 9 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/test/java/university/dao/crud/CRUDQuerySubjectTest.java]: fixed

File: NagorniyAleksandr/SQLTask/README.md (hunk 3 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/README.md]: done

File: NagorniyAleksandr/SQLTask/src/main/java/university/service/ServiceImpl.java (hunk 8 above)
Comment (NagorniyAleksandr) [reply NagorniyAleksandr/SQLTask/src/main/java/university/service/ServiceImpl.java:13]: done

File: NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java (hunk 5 above)
Comment (presly808) [reply NagorniyAleksandr/SQLTask/src/main/java/university/dao/QueryCreatorImpl.java]: ok) i dont know that is business logic

**Comment by presly808** [comment]
good job

**Pull request merged**